### PR TITLE
add function explain_statement [BF-1217]

### PR DESCRIPTION
**Pull request by carobme**
This adds a function taking a query as an argument and returning its
execution plan as JSON. Needs SECURITY DEFINER because otherwise the
user invoking this would need SELECT privileges for every table part
of the plan. Search path intentionally not set, because the user search
paths needs to be respected (otherwise statements would need to be
modified, i.e. schema-prefixed).

### CODE REVIEW

File: sql/aiven_extras.sql [hunk 1]
```diff
@@ -496,6 +496,22 @@ END;
...
 
+DROP FUNCTION IF EXISTS aiven_extras.explain_statement(TEXT);
+CREATE FUNCTION aiven_extras.explain_statement(
+    arg_query TEXT,
+    OUT execution_plan JSON
+)
+RETURNS SETOF JSON
+RETURNS NULL ON NULL INPUT
+LANGUAGE plpgsql
+SECURITY DEFINER
```
Comment (kmichel-aiven) [inline sql/aiven_extras.sql:507]: search path is missing.

**Review by kmichel-aiven — DISMISSED** [review]

File: sql/aiven_extras.sql (hunk 1 above)
Comment (carobme) [reply sql/aiven_extras.sql:507]: Search path can't be set here, see commit message.

File: sql/aiven_extras.sql [hunk 2]
```diff
@@ -496,6 +496,22 @@ END;
...
+AS $$
+BEGIN
+    RETURN QUERY EXECUTE 'EXPLAIN (FORMAT JSON) ' OPERATOR(pg_catalog.||) arg_query;
```
Comment (jankatins) [inline sql/aiven_extras.sql:510]: hasn't this a SQL injection problem? 

```SQL
SELECT aiven_extras.explain_statement($$select 1; ALTER USER jan WITH SUPERUSER; select 1 :: JSONB$$);
```

Comment (carobme) [reply sql/aiven_extras.sql:510]: That was my initial fear as well but at least with trivial examples I wasn't able to inject something:

```
=# SELECT aiven_extras.explain_statement('SELECT 1; DROP TABLE t1');
ERROR:  42P11: cannot open multi-query plan as cursor
CONTEXT:  PL/pgSQL function aiven_extras.explain_statement(text) line 3 at RETURN QUERY
LOCATION:  SPI_cursor_open_internal, spi.c:1335
```

But I'm not sure if there are ways around that, let me do a bit of investigation to be sure.

Comment (jankatins) [reply sql/aiven_extras.sql:510]: This seems to suggest that only a single statement can be used with `EXECUTE`: https://dba.stackexchange.com/a/308635/118184 although my short glance at https://www.postgresql.org/docs/current/plpgsql-statements.html found not backup for that...

But that seems like a reasonable assumption, that only a single query can be run when you want to return such a query result... :-)

Comment (carobme) [reply sql/aiven_extras.sql:510]: I had the same assumption, though I'm a bit unsure if we really can count on that behavior.

@rdunklau Maybe you can give this another pair of eyes?

P.S.: Nice seeing you around, Jan!

Comment (carobme) [reply sql/aiven_extras.sql:510]: Having some discussions about this, moving to draft for now.

File: sql/aiven_extras.sql [hunk 3]
```diff
@@ -496,6 +496,29 @@ END;
...
 
+DROP FUNCTION IF EXISTS aiven_extras.explain_statement(TEXT);
+CREATE FUNCTION aiven_extras.explain_statement(
+    arg_query TEXT,
+    OUT execution_plan JSON
+)
+RETURNS SETOF JSON
+RETURNS NULL ON NULL INPUT
+LANGUAGE plpgsql
+SECURITY DEFINER
```
Comment (jankatins) [inline sql/aiven_extras.sql:507]: I would expect the explanation from the commit message here as a comment to not let the next reviewer stumble over there missing search path.

**Comment by jankatins** [comment]
✅

**Comment by carobme** [comment]
Changed to use an explicit cursor after some discussion to not only get `ERROR:  42P11: cannot open multi-query plan as cursor` as an accidental side effect.

**Review by kmichel-aiven — APPROVED** [review]

**Pull request merged**